### PR TITLE
#350 ApplicationPage with previous and next buttons

### LINKED ISSUE #350: Create NextButton and PreviousButton and use revalidation logic
Should we create a separated component to store both and one more specific to deal with validation logic for NextButton?

### PR DESCRIPTION
**Pull request by nmadruga**
Fixes #350 

### Branched from
- ~~#375~~ MERGED

### Changes
- Create new `Navigation` component to display **Previous** and **Next** button along with **Review & Summary** (to be implemented in #365) in a raised (using Sticky) component on the bottom of the current page.
- Uses `revalidateMethod` on the Next button handler to catch any problems in the current page (using the same approach from ProgressBar - that triggers a revalidation and calls the method received once it's done).

### Unrelated changes
- Add TODOs on various components/containers to be destroyed once Re-structure for Applications is completed - pretty soon.

### CODE REVIEW

File: src/components/Application/Navigation.tsx
Comment (andreievg) [inline src/components/Application/Navigation.tsx:33]: these ones return previous or next section || null, so they can be named as previousSeciton and nextSection and used in getPreviousSecitonPage or getNextSectionPage

File: src/components/Application/Navigation.tsx
Comment (andreievg) [inline src/components/Application/Navigation.tsx:44]: as per earlier comment hasPreviousSection can be renamed to previousSection and used here

**Review by andreievg — CHANGES_REQUESTED** [review]
Logically look good, just a small simplifications suggestion.

Happy to adjust UI a bit later

File: src/components/Application/Navigation.tsx
Comment (nmadruga) [reply src/components/Application/Navigation.tsx:33]: Yep, good point.

Comment (nmadruga) [reply src/components/Application/Navigation.tsx]: Done: 7de0277

File: src/components/Application/Navigation.tsx
Comment (nmadruga) [reply src/components/Application/Navigation.tsx]: Done. Thanks : )

File: src/components/Application/Navigation.tsx
Comment (andreievg) [inline src/components/Application/Navigation.tsx:30]: hmm wonder why these need to be a method ?

**Review by andreievg — COMMENTED** [review]
Looking pretty good, will approve once tested (would need home page PR merged first, and this one updated from master, I think the polish revalidation logic is not in the home page PR)

File: src/components/Application/Navigation.tsx
Comment (nmadruga) [reply src/components/Application/Navigation.tsx:30]: Just to check before accessing the first one from the returned array (of next sections). Do you think it could be done in one line?

Comment (andreievg) [reply src/components/Application/Navigation.tsx:30]: could just be 
```JS
const nextSections = Object.values(sections)
      .filter(({ details: { index } }) => index > currentSectionDetails.index)
      .sort(({ details: { index: aIndex } }, { details: { index: bIndex } }) => aIndex - bIndex)
const nextSection = nextSections[0].details
      
```

**Comment by CarlosNZ** [comment]
One quick comment: in "non-linear" applications, pressing "Next" shouldn't do a re-validate, it should just go to the next page.

File: src/components/Application/Navigation.tsx
Comment (CarlosNZ) [inline src/components/Application/Navigation.tsx:104]: I think we should avoid using `float` for layout most of the time. It's an old-fashioned way of doing it before Flexbox was a thing, and things that are floated right don't wrap properly on mobile:
Screen Shot 2021-03-04 at 12 59 25 PM

I think you have to use Grid to lay this out properly, or maybe manually do flex box. Have a look at the Pagination component for one way of doing it (Yes, I know there is a `floated` in there, but it's actually no longer needed -- it looks the same if you remove the floats)

File: src/components/Application/Navigation.tsx
Comment (CarlosNZ) [inline src/components/Application/Navigation.tsx:72]: You don't need any parameter here, I think. Can just do:
`const previousButtonHandler = () => {`

File: src/components/Application/Navigation.tsx
Comment (CarlosNZ) [inline src/components/Application/Navigation.tsx:77]: Same -- no param required

File: src/components/Application/Navigation.tsx
Comment (CarlosNZ) [inline src/components/Application/Navigation.tsx:65]: I feel like these four methods could be combined into a single generic method that can go backwards and forwards and the input parameter specifies by how much. 

To go forward a page, you'd call: `const nextSectionPage = getSectionPage(1)`
And to go back you'd call: `const prevSectionPage = getSectionPage(-1)`

And the `getSectionPage` method would basically do:
- flatten all pages to a single, sorted array of SectionAndPage objects (i.e. has sectionCode and pageNum)
- infer the index in this array that the current page is
- Add the input param to the currentPageIndex -- if greater than .length, set newIndex = .length; if less than 0, newIndex = 0.
- return the sectionPage details of the object at the newIndex (code and page num)

(This would also work for jumping multiple pages, not that we need it, but: `getSectionPage(3)` for example

File: src/components/Application/Navigation.tsx
Comment (CarlosNZ) [inline src/components/Application/Navigation.tsx:81]: Need to check for `isLinear`, and only requestRevalidation if true.

**Review by CarlosNZ — CHANGES_REQUESTED** [review]
Not sure about the layout for this -- what design are you basing it on?

Cheers.

**Comment by nmadruga** [comment]
> One quick comment: in "non-linear" applications, pressing "Next" shouldn't do a re-validate, it should just go to the next page.

Nice catch! I thought about it the other day after finishing the work and forgot to update. Thank you!

File: src/components/Application/Navigation.tsx
Comment (nmadruga) [reply src/components/Application/Navigation.tsx:30]: Hmmm ok, there is also the check for null required there. But I'll do this change, it would be better to not have methods indeed. Thanks

Comment (nmadruga) [reply src/components/Application/Navigation.tsx]: Done: 4bd56a7

**Comment by nmadruga** [comment]
> Not sure about the layout for this -- what design are you basing it on?

The idea to display the **Previous** & **Next** buttons in the same component as **Review & Submit** button was defined in one of our meeting last week - mentioned by @andreievg. Now those lines around the buttons was just some layout that I don't have a simple way to remove and continue using the Segment.Group... So I was going to leave for improvements in the App layout epic.

File: src/containers/Application/ApplicationPage.tsx
Comment (nmadruga) [reply src/containers/Application/ApplicationPage.tsx]: No, that's how it was before (but only for the **Review & Submit** button) to display on top of the current page - at all times. See how it looks when I remove this from the Segment.Group:
Screen Shot 2021-03-04 at 1 55 37 PM

Continue with the Navigation inside looks like this:
Screen Shot 2021-03-04 at 1 53 08 PM

File: src/containers/Application/ApplicationHome.tsx
Comment (nmadruga) [reply src/containers/Application/ApplicationHome.tsx:44]: Alright. Good point, I think I have noticed this in other places - When moving from a page to the list. Not sure where though...

Comment (nmadruga) [reply src/containers/Application/ApplicationHome.tsx]: Done: cf1a450

File: src/components/Application/Navigation.tsx
Comment (nmadruga) [reply src/components/Application/Navigation.tsx:70]: But then instead of `onClick={previousButtonHandler}` I would need to do `onClick={() => previousButtonHandler}`. Preference?

File: src/components/Application/Navigation.tsx
Comment (nmadruga) [reply src/components/Application/Navigation.tsx:84]: Done: 3aac044

File: src/components/Application/Navigation.tsx
Comment (nmadruga) [reply src/components/Application/Navigation.tsx]: Hmmm I see your point, but really think it's ok since it's working correctly now and easy to read. And because no other use for this inside this component - jumping page for example. If required in the future we can change to that.

File: src/components/Application/Navigation.tsx
Comment (nmadruga) [reply src/components/Application/Navigation.tsx:107]: Yea, good point! Will have a look after the meeting : )

File: src/components/Application/Navigation.tsx
Comment (CarlosNZ) [reply src/components/Application/Navigation.tsx:70]: You don't have to do that, it works just fine with no parameter and your original `onClick`.

Comment (CarlosNZ) [reply src/components/Application/Navigation.tsx:70]: >  You only need to do the other type of `onClick` when you need to pass a parameter from the onClick

Speaking of which -- I think the nicest way would be to have a single handler function and you just send a parameter (the page offset -- see below comment) to tell it whether to go back or fowards.

So next button would have `onClick={() => navButtonHandler(1)}` and back button would have `onClick={() => navButtonHandler(-1)}`

File: src/components/Application/Navigation.tsx
Comment (nmadruga) [reply src/components/Application/Navigation.tsx:107]: Hi! I have looked in `Pagination` and changed to use Grid, but I can still see some not beautiful UI on mobile... I'm calling out for a good to have in App layout issue. But have changed to be using Grid, so hopefully better than before.

On mobile:
Screen Shot 2021-03-04 at 3 47 26 PM

On desktop:
Screen Shot 2021-03-04 at 3 47 38 PM

File: src/components/Application/Navigation.tsx
Comment (nmadruga) [reply src/components/Application/Navigation.tsx:70]: I think that's good for more generic buttons, but here the Previous is very simple compared with Next, and I choose to just do the special validation on the nextButtonHandler

Comment (CarlosNZ) [reply src/components/Application/Navigation.tsx:70]: Sure, but you still don't need the `_:any` parameters in `const previousButtonHandler = () => {...`

**Review by andreievg — APPROVED** [review]
LGTM

File: src/components/Application/Navigation.tsx
Comment (CarlosNZ) [reply src/components/Application/Navigation.tsx]: Sure, I'm not a fan of all the duplicated code -- but @andreievg 's approved and it works so carry on 👍

File: src/components/Application/Navigation.tsx
Comment (nmadruga) [reply src/components/Application/Navigation.tsx:70]: Ok. Great. I'll do this change now. : )

File: src/components/Application/Navigation.tsx
Comment (nmadruga) [reply src/components/Application/Navigation.tsx]: Yea, me neither, I would prefer to combine usually, and I see this solution as a possible one. I just think the very generic option wouldn't be as good since the Previous should get any validation at all

Comment (CarlosNZ) [reply src/components/Application/Navigation.tsx]: Sorry, couldn't resist having a quick play to see what the difference is. Check out branch `350-go-to-page-handler` (branched off this one) to see the alternative. Feel free to ignore, just wanted to see how it compared. 

Otherwise, merge away!

**Pull request merged**
